Filter invalid fields taken from summaryFields

lekoala opened the pull request:
This allows to prevent common errors like a summary_fields with Member.Title that would break some queries if no searchable_fields are provided.
While this PR fixes the issue, I think that the "relObject" probably deserve an additional parameter (like $dbFieldsOnly) and return only actual db fields. It could therefore be used in other places too. Or maybe have another method called relObjectField that would act as a simplified version of relObject.

More context here:
https://github.com/silverstripe/silverstripe-framework/issues/8792

emteknetnz commented on src/ORM/DataObject.php:
```diff
@@ -3735,6 +3735,11 @@ public function searchableFields()
                         $fields[] = $name;
                     } elseif ($this->relObject($spec)) {
+                        // Field does not always exist as a real db field (e.g. "Member.Title")
+                        $parts = explode(".", $spec);
+                        if (count($parts) == 2 && !$this->relObject($parts[0])->hasDatabaseField($parts[1])) {
+                            continue;
+                        }
```
```suggestion
                        $parts = explode('.', $spec);
                        if (count($parts) === 2) {
                            $relObject = $this->relObject($parts[0]);
                            if (!$relObject || !$relObject->hasDatabaseField($parts[1])) {
                                continue;
                            }
                        }

```

emteknetnz commented on src/ORM/DataObject.php:
```diff
@@ -3735,6 +3735,11 @@ public function searchableFields()
                     if ($schema->fieldSpec($this, $name)) {
                         $fields[] = $name;
                     } elseif ($this->relObject($spec)) {
+                        // Field does not always exist as a real db field (e.g. "Member.Title")
```
This is a bit confusing

Should this instead be along the lines of

"This is usually a real field i.e. "Member.Title", though it may be some thing else e.g. <insert example>

emteknetnz reviewed (changes requested):
Could you also add a unit test please

GuySartorelli commented:
only have limited time and ressources to invest on these matters. While I don't mind sharing what I had to fix anyway, I expect the silverstripe team to also have it's part. I think it's should be easy enough to take it from there if someone is already doing some unit tests for other stuff to also fix this one at the same time.

Understandable. I was simply following up on Steve's request since you hadn't responded. We also have limited time and resources, so I hope you can understand that adding unit tests to every community PR would be a massive undertaking.

We prefer to first ask the contributer to add tests, and if they can't (such as in this case due to time constraints) then we need to make a judgment call as to whether to add our own tests (either in the same PR or to merge and add a separate tests PR) or whether to leave it for when we have a bit more capacity. Either way that then requires someone else to review the new tests, which means even more resource is taken away from our regular sprint work.

Hopefully with that context you can understand why I followed up on Steve's request first. 😁 It wasn't at all my intention to try to put all the pressure on you so apologies if that's how it felt.

lekoala commented:
@GuySartorelli no worries, it's just that I'll probably don't do follow up on this anytime soon :)

lekoala commented:
so i'm reviving this issue since i realized i lost some custom code i made before the 4.12 update and ran into the issue today again :) 
I've updated the comment
added some basic elements to the test (didn't try locally, but i think the idea is pretty clear)

lekoala commented:
With that being said, i think it would make a lot more sense to fix the overall issue by making relObject able to distinguish between real db fields (from schema) and getters
maybe something do squeeze in for cms 5 release?

GuySartorelli commented:
This is ultimately a duplicate of the PR that ended up fixing https://github.com/silverstripe/silverstripe-framework/issues/10761 so I'll close it.
